File: android/app/src/main/java/com/freebuff/mobile/ReconnectController.kt
package com.freebuff.mobile

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.os.Handler
import android.os.Looper
import java.io.IOException
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import kotlin.math.min
import kotlin.random.Random

class ReconnectController(
    context: Context,
    private val sessionStore: SecureSessionStore,
    private val listener: (ConnectionState, String, PairingSession?) -> Unit,
) {
    private val connectivity = context.getSystemService(ConnectivityManager::class.java)
    private val executor = Executors.newSingleThreadScheduledExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())
    private val networkCallback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            scheduleConnect(immediate = true)
        }

        override fun onLost(network: Network) {
            if (!manualDisconnect) emit(ConnectionState.OFFLINE, "Network unavailable", sessionStore.load())
        }
    }

    @Volatile
    private var manualDisconnect = false
    private var retryAttempt = 0
    private var scheduled: ScheduledFuture<*>? = null
    private var refreshTimer: ScheduledFuture<*>? = null
    private var started = false

    fun start() {
        if (started) return
        started = true
        manualDisconnect = false
        runCatching { connectivity.registerDefaultNetworkCallback(networkCallback) }
        scheduleConnect(immediate = true)
    }

    fun onResume() {
        if (!started) start() else if (!manualDisconnect) scheduleConnect(immediate = true)
    }

    fun disconnect(clearSession: Boolean) {
        manualDisconnect = true
        scheduled?.cancel(false)
        refreshTimer?.cancel(false)
        refreshTimer = null
        if (clearSession) sessionStore.clear()
        emit(
            if (clearSession) ConnectionState.UNPAIRED else ConnectionState.DISCONNECTED,
            if (clearSession) "Pairing removed" else "Disconnected by user",
            sessionStore.load(),
        )
    }

    fun reconnect() {
        manualDisconnect = false
        retryAttempt = 0
        scheduleConnect(immediate = true)
    }

    fun close() {
        runCatching { connectivity.unregisterNetworkCallback(networkCallback) }
        scheduled?.cancel(true)
        refreshTimer?.cancel(true)
        executor.shutdownNow()
        started = false
    }

    private fun scheduleConnect(immediate: Boolean) {
        if (manualDisconnect || !started) return
        scheduled?.cancel(false)
        val delay = if (immediate) 0L else retryDelayMs()
        scheduled = executor.schedule({ connectOnce() }, delay, TimeUnit.MILLISECONDS)
    }

    private fun connectOnce() {
        if (manualDisconnect) return
        val stored = sessionStore.load()
        if (stored == null) {
            emit(ConnectionState.UNPAIRED, "Scan a pairing QR code", null)
            return
        }

        val reconnecting = retryAttempt > 0
        emit(
            if (reconnecting) ConnectionState.RECONNECTING else ConnectionState.CONNECTING,
            if (reconnecting) "Retrying gateway connection" else "Connecting to gateway",
            stored,
        )

        try {
            val refreshed = PairingApi(stored.gatewayBaseUrl).refresh(stored)
            sessionStore.save(refreshed)
            retryAttempt = 0
            scheduleSessionRefresh(refreshed)
            emit(ConnectionState.CONNECTED, "Gateway authenticated", refreshed)
        } catch (error: GatewayApiException) {
            if (error.status == 401 || error.status == 403) {
                sessionStore.clear()
                refreshTimer?.cancel(false)
                refreshTimer = null
                emit(ConnectionState.PAIRING_REQUIRED, "Pairing expired or revoked", null)
            } else {
                scheduleRetry(stored, error.message ?: "Gateway error")
            }
        } catch (error: IOException) {
            scheduleRetry(stored, "Waiting for network")
        } catch (error: Exception) {
            scheduleRetry(stored, error.message ?: "Connection failed")
        }
    }

    private fun scheduleSessionRefresh(session: PairingSession) {
        refreshTimer?.cancel(false)
        val expiresAt = runCatching { java.time.Instant.parse(session.accessTokenExpiresAt).toEpochMilli() }
            .getOrElse { System.currentTimeMillis() + 10 * 60_000L }
        val delay = (expiresAt - System.currentTimeMillis() - 60_000L)
            .coerceIn(30_000L, 10 * 60_000L)
        refreshTimer = executor.schedule({
            refreshTimer = null
            if (!manualDisconnect) connectOnce()
        }, delay, TimeUnit.MILLISECONDS)
    }

    private fun scheduleRetry(session: PairingSession, detail: String) {
        retryAttempt += 1
        emit(ConnectionState.RECONNECTING, detail, session)
        scheduleConnect(immediate = false)
    }

    private fun retryDelayMs(): Long {
        val exponent = min(retryAttempt - 1, 6)
        val base = min(60_000L, 1_000L shl exponent)
        val jitter = Random.nextDouble(0.8, 1.2)
        return (base * jitter).toLong()
    }

    private fun emit(state: ConnectionState, detail: String, session: PairingSession?) {
        mainHandler.post { listener(state, detail, session) }
    }
}

File: ios/FreebuffGate/ReconnectController.swift
import Foundation
import Network

class ReconnectController {
    typealias Listener = (ConnectionState, String, PairingSession?) -> Void

    private let sessionStore: SecureSessionStore
    private let listener: Listener
    private let queue = DispatchQueue(label: "com.freebuff.gate.reconnect")
    private let monitor: NWPathMonitor

    private var manualDisconnect = false
    private var retryAttempt = 0
    private var started = false
    private var pendingConnect: DispatchWorkItem?
    private var refreshTask: Task<Void, Never>?

    init(sessionStore: SecureSessionStore, listener: @escaping Listener) {
        self.sessionStore = sessionStore
        self.listener = listener
        self.monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self else { return }
            if path.status == .satisfied {
                self.scheduleConnect(immediate: true)
            } else if !self.manualDisconnect {
                self.emit(.offline, "Network unavailable", self.sessionStore.load())
            }
        }
    }

    func start() {
        if started { return }
        started = true
        manualDisconnect = false
        monitor.start(queue: queue)
        scheduleConnect(immediate: true)
    }

    func onResume() {
        if !started { start() } else if !manualDisconnect {
            scheduleConnect(immediate: true)
        }
    }

    func disconnect(clearSession: Bool) {
        manualDisconnect = true
        pendingConnect?.cancel()
        pendingConnect = nil
        refreshTask?.cancel()
        refreshTask = nil
        if clearSession { sessionStore.clear() }
        emit(
            clearSession ? .unpaired : .disconnected,
            clearSession ? "Pairing removed" : "Disconnected by user",
            sessionStore.load()
        )
    }

    func reconnect() {
        manualDisconnect = false
        retryAttempt = 0
        scheduleConnect(immediate: true)
    }

    func close() {
        monitor.cancel()
        pendingConnect?.cancel()
        refreshTask?.cancel()
        started = false
    }

    private func scheduleConnect(immediate: Bool) {
        if manualDisconnect || !started { return }
        pendingConnect?.cancel()
        let delay = immediate ? 0.0 : retryDelayMs()
        let work = DispatchWorkItem { [weak self] in self?.connectOnce() }
        pendingConnect = work
        queue.asyncAfter(deadline: .now() + delay / 1000.0, execute: work)
    }

    private func connectOnce() {
        if manualDisconnect { return }
        let stored = sessionStore.load()
        if stored == nil {
            emit(.unpaired, "Scan a pairing QR code", nil)
            return
        }
        let reconnecting = retryAttempt > 0
        emit(
            reconnecting ? .reconnecting : .connecting,
            reconnecting ? "Retrying gateway connection" : "Connecting to gateway",
            stored
        )
        Task {
            do {
                let refreshed = try await PairingApi(rawBaseUrl: stored.gatewayBaseUrl).refresh(session: stored)
                try sessionStore.save(session: refreshed)
                retryAttempt = 0
                scheduleSessionRefresh(session: refreshed)
                emit(.connected, "Gateway authenticated", refreshed)
            } catch let error as PairingError {
                switch error {
                case .http(let status, _) where status == 401 || status == 403:
                    sessionStore.clear()
                    refreshTask?.cancel()
                    refreshTask = nil
                    emit(.pairingRequired, "Pairing expired or revoked", nil)
                default:
                    scheduleRetry(stored, detail: error.localizedDescription)
                }
            } catch {
                scheduleRetry(stored, detail: "Waiting for network")
            }
        }
    }

    private func scheduleSessionRefresh(session: PairingSession) {
        refreshTask?.cancel()
        let expiresAt = ISO8601DateFormatter().date(from: session.accessTokenExpiresAt)?.timeIntervalSince1970
            ?? (Date().timeIntervalSince1970 + 10 * 60)
        let now = Date().timeIntervalSince1970
        let delay = max(30, min(600, expiresAt - now - 60))
        refreshTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            guard !Task.isCancelled else { return }
            self?.refreshTask = nil
            if !(self?.manualDisconnect ?? true) {
                self?.connectOnce()
            }
        }
    }

    private func scheduleRetry(_ session: PairingSession, detail: String) {
        retryAttempt += 1
        emit(.reconnecting, detail, session)
        scheduleConnect(immediate: false)
    }

    private func retryDelayMs() -> Double {
        let exponent = min(retryAttempt - 1, 6)
        let base = min(60_000.0, 1_000.0 * pow(2.0, Double(exponent)))
        let jitter = Double.random(in: 0.8...1.2)
        return base * jitter
    }

    private func emit(_ state: ConnectionState, _ detail: String, _ session: PairingSession?) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.listener(state, detail, session)
        }
    }
}
